File: klinara-ios/klinara-ios/Services/Auth/MockAuthService.swift
import Foundation

/// Arayüzü sürmek için gereken senaryolar.
///
/// Her biri gerçek bir sunucu durumuna karşılık gelir; hepsi
/// `HomePlaceholderView` içindeki geliştirici menüsünden değiştirilebilir.
enum MockScenario: String, CaseIterable, Identifiable, Sendable {
    case happyPasskey = "Passkey ile tek dokunuş"
    case passwordOnly = "Telefon + parola"
    case passwordThenTotp = "Parola + TOTP"
    case mfaRequiredNotConfigured = "2FA zorunlu, kurulmamış"
    case multiTenant = "Birden çok klinik"
    case multiBranch = "Birden çok şube"
    case unverifiedPhone = "Telefon doğrulanmamış"
    case wrongPassword = "Hatalı bilgi"
    case accountLocked = "Hesap kilitli"
    case rateLimited = "Hız sınırı"
    case networkError = "Bağlantı hatası"

    var id: String { rawValue }
}

/// Sahte kimlik servisi.
///
/// Yanıtları Swift struct'ı olarak **kurmaz**, gerçek JSON üretip
/// `JSONDecoder` ile çözer. Böylece mock, canlı servise geçildiğinde
/// ortaya çıkacak bir çözümleme hatasını şimdiden yakalar.
final class MockAuthService: AuthService, @unchecked Sendable {

    private let lock = NSLock()
    private var _scenario: MockScenario
    private var _phoneVerified: Bool

    init(scenario: MockScenario = .passwordThenTotp) {
        _scenario = scenario
        _phoneVerified = scenario != .unverifiedPhone
    }

    /// `NSLock` doğrudan async bağlamdan çağrılamaz (Swift 6'da hata).
    /// Kilit senkron bir yardımcının içinde kalır.
    private func withLock<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }

    var scenario: MockScenario {
        get { withLock { _scenario } }
        set {
            withLock {
                _scenario = newValue
                _phoneVerified = newValue != .unverifiedPhone
            }
        }
    }

    // MARK: Yardımcılar

    /// Canlı istemciyle BİREBİR aynı çözücü — bkz. ``KlinaraCoding``.
    private static let decoder = KlinaraCoding.decoder()

    private func decode<T: Decodable>(_ json: String) throws -> T {
        do {
            return try Self.decoder.decode(T.self, from: Data(json.utf8))
        } catch {
            throw AuthError.malformedResponse(String(describing: error))
        }
    }

    /// Yükleme durumlarının gerçekten görünür olması için ağ gecikmesi.
    private func simulateLatency(_ seconds: Double = 0.7) async {
        try? await Task.sleep(for: .seconds(seconds))
    }

    private func problem(_ code: APIErrorCode, status: Int, title: String) -> AuthError {
        .problem(ProblemDetails(
            code: code,
            title: title,
            detail: nil,
            status: status,
            requestId: "01JQMOCK\(Int.random(in: 1000...9999))"
        ))
    }

    /// Senaryonun giriş denemesini baştan reddedip reddetmediği.
    private func failureForCurrentScenario() -> AuthError? {
        switch scenario {
        case .wrongPassword:
            return problem(.invalidCredentials, status: 401, title: "Giriş başarısız")
        case .accountLocked:
            return problem(.accountLocked, status: 423, title: "Hesap kilitli")
        case .rateLimited:
            return problem(.rateLimited, status: 429, title: "Çok fazla deneme")
        case .networkError:
            return .network
        default:
            return nil
        }
    }

    // MARK: - Giriş

    func login(
        email: String?,
        phone: String?,
        password: String,
        deviceLabel: String
    ) async throws -> LoginResponse {
        await simulateLatency()
        if let failure = failureForCurrentScenario() { throw failure }

        switch scenario {
        case .multiTenant:
            return try decode(Fixtures.tenantSelectionRequired)
        case .passwordThenTotp:
            return try decode(Fixtures.mfaRequired(configured: true))
        case .mfaRequiredNotConfigured:
            return try decode(Fixtures.mfaRequired(configured: false))
        default:
            return try decode(Fixtures.authenticated)
        }
    }

    func selectTenant(challengeToken: String, tenantId: String) async throws -> LoginResponse {
        await simulateLatency(0.5)
        return try decode(Fixtures.authenticated)
    }

    func verifyMfa(challengeToken: String, code: String, deviceLabel: String) async throws -> LoginResponse {
        await simulateLatency(0.5)
        // Yedek kodlar tire içerir; TOTP daima 6 hanedir.
        let isBackupCode = code.contains("-")
        guard isBackupCode || code.count == 6 else {
            throw problem(.mfaInvalid, status: 401, title: "Kod doğrulanamadı")
        }
        // Arayüzün hata yolunu da sürebilmek için sabit bir "yanlış" kod.
        guard code != "000000" else {
            throw problem(.mfaInvalid, status: 401, title: "Kod doğrulanamadı")
        }
        return try decode(Fixtures.authenticated)
    }

    // MARK: - TOTP kurulumu

    func totpSetup(challengeToken: String) async throws -> TotpSetup {
        await simulateLatency(0.4)
        return try decode(Fixtures.totpSetup)
    }

    func totpEnable(challengeToken: String, code: String) async throws -> [String] {
        await simulateLatency(0.5)
        guard code.count == 6, code != "000000" else {
            throw problem(.mfaInvalid, status: 401, title: "Kod doğrulanamadı")
        }
        return Fixtures.backupCodes
    }

    // MARK: - Passkey

    func passkeyAssertionOptions() async throws -> PasskeyOptions {
        await simulateLatency(0.3)
        return PasskeyOptions(
            challenge: Data((0..<32).map { _ in UInt8.random(in: 0...255) }),
            relyingPartyIdentifier: "klinara.app",
            userID: nil,
            userName: nil
        )
    }

    func passkeyVerify(_ assertion: PasskeyAssertion, deviceLabel: String) async throws -> LoginResponse {
        await simulateLatency(0.4)
        if case .networkError = scenario { throw AuthError.network }
        return try decode(Fixtures.authenticated)
    }

    func passkeyRegistrationOptions() async throws -> PasskeyOptions {
        await simulateLatency(0.3)
        return PasskeyOptions(
            challenge: Data((0..<32).map { _ in UInt8.random(in: 0...255) }),
            relyingPartyIdentifier: "klinara.app",
            userID: Data("mock-user".utf8),
            userName: "+905321234567"
        )
    }

    func registerPasskey(_ registration: PasskeyRegistration, deviceLabel: String) async throws {
        await simulateLatency(0.5)
    }

    // MARK: - Oturum sonrası

    func me() async throws -> MeResponse {
        await simulateLatency(0.3)
        let (verified, multipleBranches) = withLock { (_phoneVerified, _scenario == .multiBranch) }
        return try decode(Fixtures.me(phoneVerified: verified, multipleBranches: multipleBranches))
    }

    func branches() async throws -> [BranchSummary] {
        await simulateLatency(0.3)
        let json = scenario == .multiBranch ? Fixtures.branchesMany : Fixtures.branchesOne
        return try decode(json)
    }

    // MARK: - Telefon doğrulama

    func startPhoneVerification(phone: String) async throws -> PhoneVerificationStarted {
        await simulateLatency(0.5)
        let expiry = KlinaraCoding.timestamp(Date().addingTimeInterval(300))
        return try decode("""
        { "phone": "\(phone)", "expiresAt": "\(expiry)", "delivered": true }
        """)
    }

    func verifyPhone(code: String) async throws -> PhoneVerified {
        await simulateLatency(0.5)
        guard code.count == 6, code != "000000" else {
            throw problem(.verificationFailed, status: 400, title: "Kod doğrulanamadı")
        }
        withLock { _phoneVerified = true }
        let now = KlinaraCoding.timestamp(Date())
        return try decode("""
        { "phone": "+905321234567", "verifiedAt": "\(now)" }
        """)
    }

    // MARK: - Parola kurtarma

    func forgotPassword(email: String) async throws {
        await simulateLatency(0.8)
        // Sunucu, e-posta kayıtlı olsun ya da olmasın aynı yanıtı verir.
        // Mock da bu sözleşmeyi taklit eder: asla hata fırlatmaz.
    }

    // MARK: - Oturum

    func refresh(refreshToken: String) async throws -> AuthTokens {
        await simulateLatency(0.3)
        let response: LoginResponse = try decode(Fixtures.authenticated)
        guard let tokens = response.tokens else {
            throw AuthError.malformedResponse("refresh yanıtında tokens yok")
        }
        return tokens
    }

    func logout() async throws {
        await simulateLatency(0.2)
    }
}

// MARK: - JSON fixture'ları

private enum Fixtures {

    static let authenticated = """
    {
      "status": "authenticated",
      "tokens": {
        "accessToken": "mock.access.token",
        "refreshToken": "mock-opaque-refresh-token",
        "tokenType": "Bearer",
        "expiresIn": 900
      },
      "tenant": { "id": "7f3d1a20-0000-4000-8000-000000000001" }
    }
    """

    static let tenantSelectionRequired = """
    {
      "status": "tenant_selection_required",
      "challengeToken": "mock.tenant-select.token",
      "tenants": [
        {
          "id": "7f3d1a20-0000-4000-8000-000000000001",
          "slug": "nisantasi-estetik",
          "name": "Nişantaşı Estetik",
          "roles": ["practitioner"]
        },
        {
          "id": "7f3d1a20-0000-4000-8000-000000000002",
          "slug": "bodrum-medikal",
          "name": "Bodrum Medikal Güzellik",
          "roles": ["manager", "practitioner"]
        }
      ]
    }
    """

    static func mfaRequired(configured: Bool) -> String {
        """
        {
          "status": "mfa_required",
          "challengeToken": "mock.mfa.token",
          "mfa": {
            "configured": \(configured),
            "methods": \(configured ? "[\"totp\", \"backup_code\"]" : "[\"totp\"]")
          }
        }
        """
    }

    static let totpSetup = """
    {
      "secret": "JBSWY3DPEHPK3PXP",
      "otpauthUri": "otpauth://totp/Klinara:ayse%40klinik.com?secret=JBSWY3DPEHPK3PXP&issuer=Klinara"
    }
    """

    static let backupCodes = [
        "4f2a-9c1e", "8b3d-2e7a", "1c9f-6d4b", "7e5a-3f8c", "2d6b-1a9e",
        "9a4c-7b2f", "5f8e-4c3d", "3b7d-8e1a", "6c2f-5a9b", "0e1d-3c7f",
    ]

    static func me(phoneVerified: Bool, multipleBranches: Bool) -> String {
        let memberships = multipleBranches
            ? """
              [
                { "id": "m1", "branchId": "b1000000-0000-4000-8000-000000000001",
                  "roleKey": "practitioner", "roleName": "Uygulayıcı" },
                { "id": "m2", "branchId": "b1000000-0000-4000-8000-000000000002",
                  "roleKey": "practitioner", "roleName": "Uygulayıcı" }
              ]
              """
            : """
              [
                { "id": "m1", "branchId": "b1000000-0000-4000-8000-000000000001",
                  "roleKey": "practitioner", "roleName": "Uygulayıcı" }
              ]
              """
        let branchIds = multipleBranches
            ? "[\"b1000000-0000-4000-8000-000000000001\", \"b1000000-0000-4000-8000-000000000002\"]"
            : "[\"b1000000-0000-4000-8000-000000000001\"]"

        return """
        {
          "user": {
            "id": "u1000000-0000-4000-8000-000000000001",
            "email": "ayse.yilmaz@klinik.com",
            "fullName": "Ayşe Yılmaz",
            "locale": "tr-TR",
            "isActive": true,
            "phone": "+905321234567",
            "phoneVerified": \(phoneVerified),
            "hasPassword": true,
            "lastLoginAt": "2026-08-26T08:14:00.000Z",
            "createdAt": "2026-05-02T09:30:00.000Z",
            "memberships": \(memberships)
          },
          "tenantId": "7f3d1a20-0000-4000-8000-000000000001",
          "roles": ["practitioner"],
          "permissions": ["appointment:read.own", "customer:read"],
          "branchIds": \(branchIds),
          "tenantWide": false
        }
        """
    }

    static let branchesOne = """
    [
      {
        "id": "b1000000-0000-4000-8000-000000000001",
        "name": "Nişantaşı",
        "timezone": "Europe/Istanbul",
        "address": "Teşvikiye Cad. No: 12, Şişli",
        "isActive": true
      }
    ]
    """

    static let branchesMany = """
    [
      {
        "id": "b1000000-0000-4000-8000-000000000001",
        "name": "Nişantaşı",
        "timezone": "Europe/Istanbul",
        "address": "Teşvikiye Cad. No: 12, Şişli",
        "isActive": true
      },
      {
        "id": "b1000000-0000-4000-8000-000000000002",
        "name": "Bağdat Caddesi",
        "timezone": "Europe/Istanbul",
        "address": "Bağdat Cad. No: 244, Kadıköy",
        "isActive": true
      }
    ]
    """
}

File: klinara-android/app/src/main/java/com/klinara/android/services/auth/MockAuthService.kt
package com.klinara.android.services.auth

import com.klinara.android.services.contracts.ApiErrorCode
import com.klinara.android.services.contracts.RolePermissions
import com.klinara.android.services.mock.Fixtures
import com.klinara.android.services.mock.MockIds
import com.klinara.android.services.mock.MockScenario
import com.klinara.android.services.networking.ApiError
import com.klinara.android.services.networking.KlinaraJson
import com.klinara.android.services.networking.ListEnvelope
import com.klinara.android.services.networking.ProblemDetails
import kotlinx.coroutines.delay
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.jsonObject
import java.time.Instant
import kotlin.random.Random

/**
 * Sunucusuz giriş akışı.
 *
 * **Yanıtlar gerçek JSON'dur ve üretim çözümleyicisiyle (`KlinaraJson`) çözülür.**
 * Elle kurulmuş nesneler döndürmek, bir sözleşme kaymasını mock modda gizlerdi;
 * böyleyken kayma mock'u da kırar.
 *
 * Gecikme birim testlerinde **tohumlanmış** rastgeledir: iOS'un tohumsuz sürümü yapı
 * gereği flaky test üretiyordu.
 */
class MockAuthService(
    var scenario: MockScenario = MockScenario.PasswordThenTotp,
    private val random: Random = Random.Default,
    private val latencyEnabled: Boolean = true,
) : AuthService {
    /** `verifyPhone` başarılı olunca true'ya döner — akış ilerleyebilsin. */
    private var phoneVerified: Boolean = scenario != MockScenario.UnverifiedPhone

    fun reset() {
        phoneVerified = scenario != MockScenario.UnverifiedPhone
    }

    // ------------------------------------------------------------------ giriş

    override suspend fun login(
        email: String?,
        phone: String?,
        password: String,
        deviceLabel: String,
    ): LoginResponse {
        simulateLatency()
        failureForCurrentScenario()?.let { throw it }

        val fixture =
            when (scenario) {
                MockScenario.MultiTenant -> "auth/login-tenant-selection.json"
                MockScenario.PasswordThenTotp -> "auth/login-mfa-configured.json"
                MockScenario.MfaRequiredNotConfigured -> "auth/login-mfa-not-configured.json"
                else -> "auth/login-authenticated.json"
            }
        return decode(fixture)
    }

    override suspend fun selectTenant(
        challengeToken: String,
        tenantId: String,
    ): LoginResponse {
        simulateLatency()
        return decode("auth/login-authenticated.json")
    }

    override suspend fun verifyMfa(
        challengeToken: String,
        code: String,
        deviceLabel: String,
    ): LoginResponse {
        simulateLatency()
        transportFailure()?.let { throw it }
        // Yedek kod tire içerir; TOTP tam altı hanedir. ALWAYS_WRONG hata yolunu sürmek için.
        val isBackupCode = code.contains('-')
        val isTotp = code.length == TOTP_LENGTH && code.all(Char::isDigit)
        if (code == ALWAYS_WRONG_CODE || (!isBackupCode && !isTotp)) {
            throw problem("auth/problem-mfa-invalid.json")
        }
        return decode("auth/login-authenticated.json")
    }

    override suspend fun totpSetup(challengeToken: String): TotpSetup {
        simulateLatency()
        return decode("auth/totp-setup.json")
    }

    override suspend fun totpEnable(
        challengeToken: String,
        code: String,
    ): List<String> {
        simulateLatency()
        if (code == ALWAYS_WRONG_CODE) throw problem("auth/problem-mfa-invalid.json")
        return KlinaraJson
            .decodeFromString<BackupCodes>(Fixtures.read("auth/backup-codes.json"))
            .backupCodes
    }

    // ------------------------------------------------------------------ profil

    override suspend fun me(): MeResponse {
        simulateLatency()
        transportFailure()?.let { throw it }

        val raw = KlinaraJson.parseToJsonElement(Fixtures.read("auth/me.json")).jsonObject
        val roleKey = scenario.roleKey
        val multipleBranches = scenario == MockScenario.MultiBranch

        // İzinler fixture'da TUTULMAZ: üretilmiş RolePermissions'tan gelir. Elle tutulan
        // bir liste iOS'ta bir kez saptı ve Faz 6 ekranları sessizce erişilemez oldu.
        val patched =
            JsonObject(
                raw.toMutableMap().apply {
                    put("roles", buildJsonArray { add(JsonPrimitive(roleKey)) })
                    put(
                        "permissions",
                        JsonArray(RolePermissions.forRole(roleKey).map(::JsonPrimitive)),
                    )
                    put("branchIds", JsonArray(branchIds(multipleBranches).map(::JsonPrimitive)))
                    put("user", patchedUser(raw["user"]!!.jsonObject, roleKey, multipleBranches))
                },
            )

        return KlinaraJson.decodeFromString(KlinaraJson.encodeToString(JsonObject.serializer(), patched))
    }

    override suspend fun branches(): List<BranchSummary> {
        simulateLatency()
        transportFailure()?.let { throw it }
        val fixture =
            if (scenario == MockScenario.MultiBranch) "auth/branches-many.json" else "auth/branches-one.json"
        return KlinaraJson
            .decodeFromString<ListEnvelope<BranchSummary>>(Fixtures.read(fixture))
            .data
    }

    // ------------------------------------------------------------------ telefon

    override suspend fun startPhoneVerification(phone: String): PhoneVerificationStarted {
        simulateLatency()
        return PhoneVerificationStarted(
            phone = phone,
            expiresAt = Instant.now().plusSeconds(CODE_TTL_SECONDS),
            delivered = true,
        )
    }

    override suspend fun verifyPhone(code: String): PhoneVerified {
        simulateLatency()
        if (code == ALWAYS_WRONG_CODE) throw problem("auth/problem-verification-failed.json")
        phoneVerified = true
        return PhoneVerified(phone = MOCK_PHONE, verifiedAt = Instant.now())
    }

    // ------------------------------------------------------------------ diğer

    /** ASLA hata fırlatmaz — sunucunun hesap sayımını önleyen sözleşmesinin aynısı. */
    override suspend fun forgotPassword(email: String) {
        simulateLatency()
    }

    override suspend fun logout() {
        simulateLatency()
    }

    override suspend fun passkeyAssertionOptions(): String = error("Passkey mock modda desteklenmiyor (A1.5)")

    override suspend fun passkeyVerify(
        responseJson: String,
        deviceLabel: String,
    ): LoginResponse = error("Passkey mock modda desteklenmiyor (A1.5)")

    override suspend fun passkeyRegistrationOptions(): String = error("Passkey mock modda desteklenmiyor (A1.5)")

    override suspend fun registerPasskey(
        responseJson: String,
        deviceLabel: String,
    ) = error("Passkey mock modda desteklenmiyor (A1.5)")

    // ------------------------------------------------------------------ iç

    /**
     * Girişe özgü hata senaryoları. `WrongPassword`, `AccountLocked` ve `RateLimited`
     * yalnız kimlik doğrulamada anlamlıdır; oturum açıldıktan sonraki çağrıları
     * etkilemezler.
     */
    private fun failureForCurrentScenario(): ApiError? =
        when (scenario) {
            MockScenario.WrongPassword -> problem("auth/problem-invalid-credentials.json")
            MockScenario.AccountLocked -> problem("auth/problem-account-locked.json")
            MockScenario.RateLimited -> problem("auth/problem-rate-limited.json")
            else -> transportFailure()
        }

    /**
     * Taşıma katmanı hatası HER çağrıyı keser.
     *
     * Yalnız `login`'i kesmek gerçekçi değildi: bir ağ kesintisi oturum geri yükleme
     * sırasındaki `me()` çağrısını da düşürür ve açılış yolunun o hataya dayanıklı
     * olduğunu ancak böyle test edebiliriz.
     */
    private fun transportFailure(): ApiError? =
        ApiError.Network().takeIf { scenario == MockScenario.NetworkError }

    private fun problem(path: String): ApiError.Problem =
        ApiError.Problem(KlinaraJson.decodeFromString<ProblemDetails>(Fixtures.read(path)))

    private inline fun <reified T> decode(path: String): T = KlinaraJson.decodeFromString(Fixtures.read(path))

    private fun branchIds(multiple: Boolean) =
        if (multiple) {
            listOf(MockIds.BRANCH_NISANTASI, MockIds.BRANCH_BODRUM)
        } else {
            listOf(MockIds.BRANCH_NISANTASI)
        }

    private fun patchedUser(
        user: JsonObject,
        roleKey: String,
        multipleBranches: Boolean,
    ): JsonObject {
        val roleName = if (roleKey == "practitioner") "Uygulayıcı" else "Şube Yöneticisi"
        val memberships =
            JsonArray(
                branchIds(multipleBranches).mapIndexed { index, branchId ->
                    JsonObject(
                        mapOf(
                            "id" to JsonPrimitive("m${index + 1}"),
                            "branchId" to JsonPrimitive(branchId),
                            "roleKey" to JsonPrimitive(roleKey),
                            "roleName" to JsonPrimitive(roleName),
                        ),
                    )
                },
            )
        return JsonObject(
            user.toMutableMap().apply {
                put("phoneVerified", JsonPrimitive(phoneVerified))
                put("memberships", memberships)
            },
        )
    }

    private suspend fun simulateLatency() {
        if (!latencyEnabled) return
        delay(random.nextLong(MIN_LATENCY_MILLIS, MAX_LATENCY_MILLIS))
    }

    @kotlinx.serialization.Serializable
    private data class BackupCodes(val backupCodes: List<String> = emptyList())

    private companion object {
        const val TOTP_LENGTH = 6

        /** Hata yolunu sürmek için sabit, her zaman yanlış kod. */
        const val ALWAYS_WRONG_CODE = "000000"

        const val CODE_TTL_SECONDS = 300L
        const val MOCK_PHONE = "+905321234567"

        const val MIN_LATENCY_MILLIS = 200L
        const val MAX_LATENCY_MILLIS = 800L
    }
}
